### PR TITLE
meta-lmp: xilinx-tools: bitstream-extraction: Add bitstream support

### PR DESCRIPTION
**Pull request by Tim-Anderson**
Signed-off-by: Tim Anderson <tim.anderson@foundries.io>

### CODE REVIEW

File: meta-lmp-bsp/conf/machine/include/ultrazed-lmp.inc [hunk 1]
```diff
@@ -55,3 +55,7 @@ PMU_FIRMWARE_DEPLOY_DIR ?= "${DEPLOY_DIR_IMAGE}"
 
 # Not building for QEMU standard Xinix build
 EXTRA_IMAGEDEPENDS_remove = "qemu-devicetrees"
+
+# bitstream support
+MACHINE_FIRMWARE_append_uz3eg-iocc = " bitstream-extraction fpga-manager-script"
```
Comment (ricardosalveti) [inline meta-lmp-bsp/conf/machine/include/ultrazed-lmp.inc:60]: Since this is machine specific, move this line to conf/machine/uz3eg-iocc.conf instead.

Also, please split the machine changes into a different commit.

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+do_compile[noexec] = "1"
+
+SYSROOT_DIRS += "${nonarch_base_libdir}/firmware"
+
+do_install() {
+
+    if [ -e ${XSCTH_WS}/${XSCTH_PROJ}_hwproj/*.bit ]; then
+        install -d ${D}/${nonarch_base_libdir}/firmware/
+        install -d ${D}/var/lib/firmware/bitstream/
```
Comment (ricardosalveti) [inline meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb:24]: Var is a special directory in our setup because it is RW, so not really managed by ostree.

As all you need is a link, it is better to create a systemd tmpfile for this recipe, like https://github.com/foundriesio/meta-lmp/blob/master/meta-lmp-base/recipes-security/softhsm/softhsm/tmpfiles.conf (but using link instead, check https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html for more info). This allows the link to be created by systemd during boot automatically, and it will keep working even if the user erases /var completely (as it will be restored on a following boot).

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        ln -s -r ${D}/var/lib/firmware/bitstream ${D}/${nonarch_base_libdir}/firmware/bitstream
+        install -Dm 0644 ${XSCTH_WS}/${XSCTH_PROJ}_hwproj/*.bit ${D}/${nonarch_base_libdir}/firmware/
+    fi
+}
+
```
Comment (ricardosalveti) [inline meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb:29]: For loading from fit you will also need a do_deploy (no needed for now, but just a heads up).

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb (hunk 2 above)
Comment (ricardosalveti) [reply meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb:24]: Actually, your recipe will have to install the link and the systemd tmpfiles configuration will have to create the directory (under var).

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/fpga-manager-script/files/0001-Correct-used-paths.patch [hunk 4]
```diff
@@ -0,0 +1,160 @@
+From 84394ab52b5eedc5c1220b7988a44db617b98691 Mon Sep 17 00:00:00 2001
+From: Tim Anderson <tim.anderson@foundries.io>
+Date: Fri, 22 Jan 2021 22:56:30 +0000
+Subject: [PATCH] Correct used paths
+
+This updates the utility to use paths that are correct for this
+distribution.
+
```
Comment (ricardosalveti) [inline meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/fpga-manager-script/files/0001-Correct-used-paths.patch:8]: While the configpath change can easily land upstream, please add the following line (to track the patch upstream status):

Upstream-Status: Inappropriate [lmp specific]

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/fpga-manager-script/fpga-manager-script_%.bbappend [hunk 5]
```diff
@@ -0,0 +1,5 @@
+FILESEXTRAPATHS_prepend := "${THISDIR}/files:"
+
+SRC_URI_append += " \
```
Comment (ricardosalveti) [inline meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/fpga-manager-script/fpga-manager-script_%.bbappend:3]: No need to use append with += (either append or +=).

**Review by ricardosalveti — CHANGES_REQUESTED** [review]
Comments inline.

File: meta-lmp-bsp/conf/machine/include/ultrazed-lmp.inc (hunk 1 above)
Comment (Tim-Anderson) [reply meta-lmp-bsp/conf/machine/include/ultrazed-lmp.inc:60]: will do

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb (hunk 2 above)
Comment (Tim-Anderson) [reply meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb:24]: the kernel driver will only load from /lib/firmware/... so the symlink has to point to someplace that is writable. and since the utility creates the /var/lib/firmware/bitstream automatically I do not think a systemd operation is strictly neccessary.

File: meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb (hunk 3 above)
Comment (Tim-Anderson) [reply meta-lmp-bsp/dynamic-layers/xilinx-tools/recipes-bsp/bitstream/bitstream-extraction_git.bb:29]: understood, I will add that when I look at the u-boot pieces

File: meta-lmp-bsp/conf/machine/uz3eg-iocc.conf [hunk 6]
```diff
@@ -12,3 +12,7 @@ HDF_MACHINE ?= "zcu102-zynqmp"
 HDF_BASE = "file://"
 HDF_PATH = "system.dsa"
 HDF_EXT = "dsa"
+
+# bitstream support
+MACHINE_FIRMWARE_append_uz3eg-iocc = " bitstream-extraction fpga-manager-script"
```
Comment (ricardosalveti) [inline meta-lmp-bsp/conf/machine/uz3eg-iocc.conf:17]: No need for append_uz3eg-iocc as this is already the board configuration file, just append works.

**Review by ricardosalveti — APPROVED** [review]
LGTM
